tests/testthat/test-shape.R

```diff
@@ -83,3 +83,23 @@ test_that("shape_broadcast_() applies recycling rules", {
     class = "vctrs_error_incompatible_type"
   )
 })
+
+test_that("can combine shaped native classes (#1290, #1329)", {
+  x <- Sys.time() + c(1, 1e6)
+  dim(x) <- c(1, 2)
+  out <- vec_c(x, x)
+
+  expect_s3_class(out, c("POSIXct", "POSIXt"))
+  expect_dim(out, c(2, 2))
+
+  y <- Sys.time() + 1:3
+  dim(y) <- c(1, 3)
+
+  expect_snapshot(error = TRUE, vec_c(x, y))
+
+  d <- structure(Sys.Date(), dim = 1)
+  expect_equal(
+    vec_rbind(data.frame(d), data.frame(d)),
+    data.frame(d = structure(rep(Sys.Date(), 2), dim = 2))
+  )
+})
```

Comment:
This isn't the best error message

```r
x <- factor(c("x", "y", "z"))
dim(x) <- c(3, 1)

y <- factor(c("x", "y", "z"))

vec_cast(x, y)
#> Error:
#> ! Can't convert `x` <factor<32af0>> to <factor<32af0>>.
#> Cannot decrease dimensions.
```

Maybe we could at least mention the dimensionality change? i.e.

```r
Can't decrease dimensionality from 2 to 1.
```

Reply:
This is now:

```
Error:
! Can't convert `x` <factor<32af0>[,1]> to <factor<32af0>>.
Can't decrease dimensionality from 2 to 1.
```